python/tests/test_prepare_release.py

```diff
@@ -0,0 +1,116 @@
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE,
+                                 '     short | recursive ) echo "Configuration of Citus 10.2devel:";;')
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE, "PACKAGE_VERSION='10.2devel'")
+        assert os.path.exists(f"{TEST_BASE_PATH}/{newly_created_sql_file}")
+        #
+        run(f"git checkout {MAIN_BRANCH}")
+        clear_branches(release_branch, resource_status, upcoming_version_branch)
+    finally:
+        clear_env()
+
+
+def clear_branches(release_branch, resource_status, upcoming_version_branch):
+    if resource_status in (ResourceStatus.RELEASE_BRANCH_LOCAL, ResourceStatus.RELEASE_BRANCH_REMOTE,
+                           ResourceStatus.UPCOMING_VERSION_LOCAL, ResourceStatus.UPCOMING_VERSION_REMOTE,
+                           ResourceStatus.PULL_REQUEST_CREATED):
+        run(f"git branch -D {release_branch} ")
+
+    if resource_status in (ResourceStatus.UPCOMING_VERSION_LOCAL, ResourceStatus.UPCOMING_VERSION_REMOTE,
+                           ResourceStatus.PULL_REQUEST_CREATED):
+        run(f"git branch -D {upcoming_version_branch} ")
+
+
+def test_patch_release():
+    initialize_env()
+    os.chdir("citus")
+    try:
+        update_release(github_token=github_token, project_name="citus", project_version="10.2.0",
+                       main_branch=MAIN_BRANCH,
+                       earliest_pr_date=datetime.strptime('2021.03.25 00:00', '%Y.%m.%d %H:%M'),
+                       exec_path=TEST_BASE_PATH, is_test=True)
+
+        release_branch, upcoming_version_branch, newly_created_sql_file, resource_status = update_release(
+            github_token=github_token, project_name="citus", project_version="10.2.1",
+            main_branch=MAIN_BRANCH,
+            earliest_pr_date=datetime.strptime('2021.03.25 00:00', '%Y.%m.%d %H:%M'),
+            exec_path=TEST_BASE_PATH, is_test=True)
+        assert file_include_line(TEST_BASE_PATH, MULTI_EXTENSION_OUT, " 10.2.1")
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE_IN, "AC_INIT([Citus], [10.2.1])")
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE, "PACKAGE_VERSION='10.2.1'")
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE, "PACKAGE_STRING='Citus 10.2.1'")
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE,
+                                 r"\`configure' configures Citus 10.2.1 to adapt to many kinds of systems.")
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE,
+                                 '     short | recursive ) echo "Configuration of Citus 10.2.1:";;')
+        assert file_include_line(TEST_BASE_PATH, CONFIGURE, "PACKAGE_VERSION='10.2.1'")
+        run(f"git checkout {MAIN_BRANCH}")
+
+        clear_branches(release_branch, resource_status, upcoming_version_branch)
+    finally:
+        clear_env()
+
+
+def clear_env():
+    if os.path.exists("../citus"):
+        os.chdir("..")
+        run("chmod -R 777 citus")
```

Comment:
why do we need that (since we are anyway removing this directory) ?

Reply:
I think it is better to clean directories and files before running the tests, not after. This way, it is easier to investigate and understand why a test fails.

Reply:
I think keeping environment clean is important if we clean at the start, there will be some residual files after test execution which is not a good practice. If we need to clean environment before test execution, we need to clean environment after the tests as well to remove any residual files which is unnecessary and increases the test duration

Reply:
I agree with Hanefi on doing the cleanup before starting to do anything.
Consider this scenario: We rely on that script does the clean-up in the end.
If we interrupt the script in the middle of its execution, then we wouldn't clean-up anything.

Other than that, I think we should create a temporary directory before doing anything.
Consider below:
```python
def clear_env():
    if os.path.exists("../citus"):
        os.chdir("..")
        # run("chmod -R 777 citus")
        run("sudo rm -rf citus")
```

This function removes `../citus` directory always, no matter if it's created by the script or not.
This is dangerous a lot since that directory might be a real directory used by the person who runs the script in their local environment.

We can chat over teams if there is anything unclear.